### DISCUSSION TITLE
Attached Properties Weird Syntax, and extra parens

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#10654] [General]
What is this weird syntax for `[Canvas.LeftProperty] = 110`?  It appears to be assigning an attached property, but it uses square brackets.  Is this a C# thing or is Avalonia doing something special here?  And why is this wrapped in parens?

Taken from https://csharp.hotexamples.com/examples/Avalonia.Controls/Canvas/-/php-canvas-class-examples.html
![image](https://user-images.githubusercontent.com/672429/224907086-5424a24e-787a-4710-a2a0-daccad31e947.png)

**Comment by rabbitism** [comment]
The syntax comes from here 
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/AvaloniaObject.cs#L86-L90

**Comment by robloo** [comment]
The square brackets are an indexer which are common in C#. It's just like a list that allows using .Items[0]. What is commonly forgotten is that an indexer doesn't have to use an integer. It can use a string or in this case an AvaloniaProperty as @rabbitism pointed out.

Keep in mind attached properties are properties registered to the control using the property system -- but these are not actual properties on the control itself. So you can't set attached properties like you can Width/Height/ClipToBounds in the code you have above. Using C# indexers provides a nice trick to work around this limitation and you can set it like any other property now, just put it in brackets using indexer syntax.

I think this trick first started with devs that wanted to write C# for UI declaration instead of XAML. So probably in the Xamarin projects at some point and then the idea was taken here in Avalonia because it's a very creative syntax to solve this problem.

**Reply by sonicsmooth** [reply]
Thanks for the explanation.  I found this page also useful.
https://docs.avaloniaui.net/docs/data-binding/binding-from-code#subscribing-to-a-property-on-any-object
